### PR TITLE
Inductive Complex

### PR DESCRIPTION
**Pull request by amrmx**
This is working code for the inductive construction of the Vietoris–Rips Complex.

### CODE REVIEW

File: ctl/vr/inductive_complex.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#include <algorithm>
+
+//BOOST
+#include <boost/graph/adjacency_list.hpp>
+
+namespace ctl {
+
+// Given a node, we loop through all its neighbors and return a vector
+// of all the neighbors that have a lower ordering.
+template<typename Graph, typename Vertex, typename Points>
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: this exists elsewhere in the library. do not duplicate functionality.

File: ctl/vr/inductive_complex.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	for ( int i = 0; i < k_cells.size(); i++) {
+	    Simplex tau(k_cells[i]);
+	    // getting the intersection of all lower neighbors
+	    std::vector< Vertex> final_neighbors;
+            std::vector< Vertex> lower_neighbors;
+	    get_lower_neighbors(g, *(tau.begin()), final_neighbors);
+	    for(simplex_iterator vi = tau.begin(); vi != tau.end(); ++vi) {
+		lower_neighbors.clear();
+		get_lower_neighbors(g, *vi, lower_neighbors);
+		std::set< Vertex> finalN(final_neighbors.begin(), final_neighbors.end());
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: Much of this is highly inefficient. A number of temporaries can be removed. Until this is resolved I won't accept the pull.

Also, I am adding the gtest unit test framework. Doing some unit testing on these algorithms would be useful.

File: ctl/vr/inductive_test.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    std::cerr << "Graph: " <<  boost::num_vertices( graph) << ", "
+	      << boost:: num_edges( graph) << std::endl;
+    std::cerr << "Graph construction: " << clock.elapsed() << std::endl;
+    clock.start();
+    Complex complex; 
+    ctl::inductive_vr(graph, complex, dimension);
+    clock.stop();
+    std::cerr << "Inductive construction: " << clock.elapsed() << std::endl;
+    std::cerr << "Number of simplices: " << complex.size() << std::endl;
+    std::cerr << "dimension of complex: " << complex.dimension() << std::endl;
```
Comment (rhl-) [inline ctl/vr/inductive_test.cpp]: the _test.cpp should really be a tool. i'm pushing some changes with many updates to the directory layout though, including moving the tools out of the ctl/ source tree..

perhaps making  unified 'vr' tool would be good, which allows you to select the algorithm on the command line would be better...

**Comment by rhl-** [comment]
I've added a number of comments. I think these changes should be made. Use a profiler like Zoom to see where bottlenecks are and remove them.

File: ctl/vr/inductive_complex.h [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                k_plus_one_cells.push_back(sigma);
+                complex.insert_open_cell(sigma);
+        }
+    }
+}
+
+// We loop through all nodes, selecting each one as the starting point for
+// constructing our simplicial complex.
+template<typename Graph, typename Complex> 
+void inductive_vr (const Graph& g, Complex& complex, std::size_t dimension) {
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: Much of this is highly inefficient. A number of temporaries can be removed. Until this is resolved I won't accept the pull.

Also, I am adding the gtest unit test framework. Doing some unit testing on these algorithms would be useful.

File: ctl/vr/inductive_complex.h [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    typedef typename Simplex::value_type Vertex;
+    typedef typename Simplex::const_iterator simplex_iterator;
+    std::vector<Simplex> k_cells;
+    std::vector<Simplex> k_plus_one_cells;
+    // adding the graph to the complex
+    add_graph_to_complex (g, complex, k_cells, k_plus_one_cells);
+    // inductive algorithm
+    for ( int k = 1; k < dimension; k++) {
+        k_cells.swap(k_plus_one_cells);
+	k_plus_one_cells.clear();
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: What is the time difference between .clear() and .erase( begin(), end() ) ?

the semantics of this are important..

File: ctl/vr/inductive_complex.h [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	for ( int i = 0; i < k_cells.size(); i++) {
+	    Simplex tau(k_cells[i]);
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: This copy is pointless, right?
const Simplex & tau = k_cells[ i];

File: ctl/vr/inductive_complex.h [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	    std::vector< Vertex> final_neighbors;
+            std::vector< Vertex> lower_neighbors;
+	    get_lower_neighbors(g, *(tau.begin()), final_neighbors);
+	    for(simplex_iterator vi = tau.begin(); vi != tau.end(); ++vi) {
+		lower_neighbors.clear();
+		get_lower_neighbors(g, *vi, lower_neighbors);
+		std::set< Vertex> finalN(final_neighbors.begin(), final_neighbors.end());
+		std::set< Vertex> lowerN(lower_neighbors.begin(), lower_neighbors.end());
+		final_neighbors.clear();
+		set_intersection(lowerN.begin(),
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: I would just do this:
for( auto & vi: tau){
     get_lower_neighbors(  g, vi, lower_neighbors);
     Simplex shared_neighbors;
     set_intersection( final_neighbors.begin(), final_neighbors.end(), 
                                lower_neighbors.begin(), lower_neighbors.end(), 
                                back_inserter( shared_neighbors));
     // then 
final_neighbors.swap( shared_neighbors);

In other words: lower_neighbors and final_neighbors don't need to be copied. If they are stored in sorted order in the vector then set_intersection would work just fine on those containers directly. 

Also, you missed an obvious optimization, if the shared_neighbors vector becomes empty, we may as well bail out early:
if( final_neighbors.empty()) { break; }

File: ctl/vr/inductive_complex.h [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    // adding the graph to the complex
+    add_graph_to_complex (g, complex, k_cells, k_plus_one_cells);
+    // inductive algorithm
+    for ( int k = 1; k < dimension; k++) {
+        k_cells.swap(k_plus_one_cells);
+	k_plus_one_cells.clear();
+	for ( int i = 0; i < k_cells.size(); i++) {
+	    Simplex tau(k_cells[i]);
+	    // getting the intersection of all lower neighbors
+	    std::vector< Vertex> final_neighbors;
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: These are really just a Simplex right? may as well just use that type..

File: ctl/vr/inductive_complex.h [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+				 lowerN.end(),
+				 finalN.begin(),
+				 finalN.end(),
+				 back_inserter(final_neighbors));
+	    }
+	    // constructing new simplices and adding them to the complex
+    	    for( int j = 0; j < final_neighbors.size(); ++j) {
+		Simplex sigma( tau);
+		sigma.insert( final_neighbors[j]); 
+		k_plus_one_cells.push_back(sigma);
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: I would try make use of emplace_back here and I would avoid the explicit loop.

e.g. you have  simplices tau and "final_neighbors" and you wish to take there union. 
You essentially want this:
k_plus_one_cells.emplace_back( union.begin(), union.end());

but: you just have the two simplices.

There are two ways to do this, you could manually write a concatenation iterator which takes the two begin(), end() pairs and turns them into one begin()/end() pair. 

e.g: http://stackoverflow.com/questions/981186/chain-iterator-for-c

OR

You could try something like const auto & union =  boost::range::join( tau, final_neighbors);

Finally, after you create the object in the k_plus_one_cells structure, then insert it into the complex...

File: ctl/vr/inductive_complex.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	    for( auto & vi: tau) {
+		if( final_neighbors.empty()) { break; }
+		lower_neighbors.clear();
+		ctl::get_lower_neighbors( g, vi, lower_neighbors);
+		std::set< Vertex> intersect;
+		set_intersection(lower_neighbors.begin(),
+				 lower_neighbors.end(),
+				 final_neighbors.begin(),
+				 final_neighbors.end(),
+				 std::inserter(intersect,intersect.begin()));
```
Comment (rhl-) [inline ctl/vr/inductive_complex.h]: Why did you change back inserter? Never insert into the front of a vector! Do you understand why?

Comment (rhl-) [inline ctl/vr/inductive_complex.h]: And the copy here is again irrelevant no?

**Comment by rhl-** [comment]
First: Resolve the merge conflicts:


Second, looking at the code, I have some comments:

final_neighbors.clear();
std::copy (intersect.begin(),intersect.end(),back_inserter(final_neighbors));

I claim this second copy is irrelevant. The point here is that intersect is just a temporary. You may as well: - Make intersect a vector and back_insert into it. - Immediately after the set_intersection() you really just need to rename intersect to final_neighbors. e.g. final_neighbors.swap( intersect) Now intersect contains some spurious data. We may now safely erase the entire container. This saves the call to std::copy and the out of cache accesses on your std::map I'm not sure that intersect.clear() is the right way though. measure the difference between that and erase(begin(), end()) i say this because .clear() may deallocate the underlying memory, in which case you would just reallocate it back via the push_back(). Which would waste time on context switches.. Also, deconstructor call on the underlying integer can perhaps be optimized away since the destructor is trivial. Although I don't know if this actually happens.

Do report some timings on the  bunny dataset for the current state and after my suggestions..
